activerecord/lib/active_record/connection_adapters/connection_specification.rb

```diff
@@ -17,6 +17,45 @@ def to_hash
         @config.merge(name: @name)
       end
 
+      class ConnectionConfigurations  #:nodoc:
+
+        def initialize(config)
+          @config = config
+          @root_level = nil
+          if url = ENV["DATABASE_URL"]
+            @database_url_config = ConnectionUrlResolver.new(url).to_hash
+          else
+            @database_url_config = nil
+          end
+        end
+
+        attr_accessor :root_level
+
+        def [](key)
+          ret = nil
+          if @root_level && c = @config[@root_level]
+            ret = c[key]
+          end
+          ret ||= @config[key]
+          if @database_url_config
+            ret = (ret || {}).merge(@database_url_config)
+          end
+          ret
+        end
+
+        def keys
+          if @root_level
+            @config[@root_level].keys
+          else
+            @config.keys
+          end
+        end
+
+        def to_hash
+          @config
+        end
+      end
```

Comment:
I still found a lot of this logic hard to read, so I took the liberty of trying to make it more concise:

```ruby
class ConnectionConfigurations #:nodoc:
  def initialize(config)
    @root_config = @config = config
    @database_url_config = resolve_database_url_config || {}
  end
 
  def scope=(scope)
    @root_config, @config = @config, (@config[scope] || {})
  end
 
  def [](key)
    @database_url_config.reverse_merge(@config[key] || @root_config[key] || {})
  end

  def keys
    @config.keys
  end

  def to_h
    @config
  end

  private
    def resolve_database_url_config
      if ENV["DATABASE_URL"]
        ConnectionUrlResolver.new(ENV["DATABASE_URL"]).to_hash
      end
    end
end
```

I'm curious why we need the `@root_config[key]` fallback in `[]`?

NOTE: `scope=` basically assumes that it's never assigned more than once, which it doesn't seem to be in library code, but it is in test code. Would it make more sense to pass it in at the constructor? Or have a `scope` method that returns a new instance of `ConnectionConfigurations`?